content/posts/using-the-redis-allocator-in-rust.md

```diff
@@ -0,0 +1,153 @@
+41  redis-server                        0x000000010ad94428 main + 1190
+...
+```
+
+It still took me a lot head scratching and experimenting to figure it out, but here's what happened:
+
+The functions of the Redis modules API are accessed via C function pointers. Instead of relying on the dynamic linker to initialize these pointers, they are initialized explicitly by Redis as part of module initialization process.
+
+As the stack trace shows, during the loading of the module we call the `CString::new` function. This standard library library function allocates memory for a string. This, in turn, calls our allocator which would then call `RedisModule_Alloc.unwrap()...` to actually perform the allocation. This causes a chicken-and-egg problem. The Redis module is not ready yet, meaning our function pointers have not yet been initialized, so we can't call the relevant API to perform the allocation.
+
+## The Solution
+
+I tried various approaches to solve this, but there seemed to be no clean way to avoid the allocation during module initialization. The second best thing would have been to use the standard allocator until the module is ready, and then switching to the custom one. However, Rust doesn't allow changing the allocator at runtime so we can't do that.
+
+What I ended up doing was to add a flag to the custom allocator that causes allocations to be passed through to the system allocator at startup. After the module initialization is complete, the flag is toggled so that further allocations are then performed via the Redis allocator. This solution still has edge cases, most importantly requiring that all previously allocated memory is freed before switching, otherwise that memory would leak. However, it's good enough for our purposes.
+
+Here is what the final code looks like:
+
+```rust
+use ...;
+use std::sync::atomic::{AtomicBool, Ordering::SeqCst};
+
+pub struct RedisAlloc;
+
+static USE_REDIS_ALLOC: AtomicBool = AtomicBool::new(false);
+
+unsafe impl GlobalAlloc for RedisAlloc {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let use_redis = USE_REDIS_ALLOC.load(SeqCst);
+        if use_redis {
+            return raw::RedisModule_Alloc.unwrap()(layout.size()) as *mut u8;
+        }
+        System.alloc(layout)
+    }
+
+    unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
+        let use_redis = USE_REDIS_ALLOC.load(SeqCst);
+        if use_redis {
+            return raw::RedisModule_Free.unwrap()(ptr as *mut c_void);
+        }
+        System.dealloc(ptr, layout);
+    }
+}
+
+pub fn use_redis_alloc() {
+    USE_REDIS_ALLOC.store(true, SeqCst);
+    eprintln!("Now using Redis allocator");
+}
+```
+
+We add a `static` flag named `USE_REDIS_ALLOC` that determine whether we should use the Redis allocator or the system one. It's important to guarantee safety when mutating static data, so we use an `AtomicBool` here that is `false` by default.
```

Comment:
Change to: We add a `static` flag named `USE_REDIS_ALLOC` that determines whether we should use the Redis allocator or the system one. It's important to guarantee safety when mutating static data, so we use an `AtomicBool` here that is `false` by default.